### ISSUE TITLE
[Question] 请教下Pi3X的PointLoss相比Pi3增加了depth_loss和rays_loss,，这是出于什么考虑呢？

### ISSUE THREAD

**Issue by buzzcut2190** [yyfz/Pi3#142]
PointLoss中的代码：

        local_pts_loss = self.criteria_local(aligned_pred_local_pts[valid_masks].float(), gt_local_pts[valid_masks].float()) * weights[valid_masks].float()[..., None]
        details['local_pts_loss'] = local_pts_loss.mean()
        pix = torch.from_numpy(get_pixel(H, W).T.reshape(H, W, 3)).to(gt_local_pts.device).float()[:, :].repeat(B, N, 1, 1, 1)
        gt_rays = torch.einsum('bnij, bnhwj -> bnhwi', torch.inverse(gt['camera_intrinsics']), pix)[..., :2]
        rays_loss = F.l1_loss(pred['xy'], gt_rays)
        final_loss += rays_loss
        details['rays_loss'] = rays_loss

        depth_loss = local_pts_loss[..., 2].mean()
        details['depth_loss'] = depth_loss
        final_loss += depth_loss

另外，从代码中看teacher模型推出的result['local_points']是深度吗？应该还不是points?

        with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
            result = self.teacher(imgs, gt['camera_poses'][teach_batch_id], gt['camera_intrinsics'][teach_batch_id])
        new_depth = result['local_points']
        pix = torch.from_numpy(get_pixel(H, W).T.reshape(H, W, 3)).to(imgs.device).float()[:, :].repeat(B, N, 1, 1, 1)
        gt_rays = torch.einsum('bnij, bnhwj -> bnhwi', torch.inverse(gt['camera_intrinsics'][teach_batch_id]), pix)[..., :2]
        pred_pts_cam = torch.cat([gt_rays * new_depth, new_depth], dim=-1)

**Comment by yyfz** [comment]
您好，teacher模型因为会输入相机参数，所以我直接让他预测depth了，没必要预测xyz，这样效果会好点。我在pi3x把点云loss分解成了ray loss和depth loss，因为ray我们总能得到dense的，但是depth很多时候是sparse的，没法全部监督，这样我们就能预测dense的ray了，这对于后面ray转为内参矩阵很友好。

**Comment by buzzcut2190** [comment]
不好意思，您说的“因为ray我们总能得到dense的，但是depth很多时候是sparse的，没法全部监督，这样我们就能预测dense的ray了”如何理解:
为什么总能得到dense ray，您是说因为gt的ray是dense的，这样通过ray_loss也能获得dense的ray吗？
dense的ray转为内参矩阵很友好是因为，整张图的投影逻辑更连贯吗？

**Comment by yyfz** [comment]
您的理解是对的，就是ray能够有dense的gt监督，内参的所有信息就在ray中，理论上不需要depth，所以完整的ray更能算出对的intrinsic. 参考https://github.com/yyfz/Pi3/blob/main/example_mm.py#L119

**Comment by buzzcut2190** [comment]
您好，打扰了，还有几个问题想再请教下：
1. gt里的sparse_depth_masks含义是什么？
2. camera_loss 中基于共视的降权reweight还没有用起来，这是实验性策略吗？
3. Teacher模型是否需要conf输出，还是说认为teacher是置信度很高的结果？
4. 还有就是能不能请教下teacher的dataset都有哪些，在选择的时候有没有什么标准呢？

**Comment by yyfz** [comment]
1. 就是depth condition要随机drop一些pixel，构造稀疏depth condition。
2. 试了没用
3. 我没加conf，感觉太麻烦了
4. 用了所有有camera和depth gt的数据
